### DISCUSSION TITLE
How to call MnistDataSet.read_data_sets

### DISCUSSION THREAD

**Discussion by pocky6** [SciSharp/TensorFlow.NET#1138] [Q&A] [answered]
Hi, 

I have a sample code which used MNIST dataset, using 
```cs
keras.datasets.mnist.load_data(); 
```
I updated drivers, cuda, cudnn, the packages to their latest version, changed OptimizerV2 to IOptimizer where needed. I get this error:
```
System.TypeInitializationException : 'The type initializer for 'Tensorflow.KerasApi' threw an exception.'
TypeLoadException : Method 'LSTMCell' in type 'Tensorflow.Keras.Layers.LayersApi' from assembly 'Tensorflow.Keras, Version=0.10.5.0, Culture=neutral, PublicKeyToken=cc7b13ffcd2ddd51' does not have an implementation.
```
I came to the docs (on readthedocs.io) which mentions the following code instead:
```cs
mnist = MnistDataSet.read_data_sets("mnist", one_hot: true);
```
But this method (read_data_sets) seems unknown from the current TF.Net source, which looks weird because it was a convenient way of getting it.
What is now the standard way of loading it ? Where is the actual updated version of the current version ?

Thank you in advance

**Comment by Wanglongzhi2001** [comment, marked as answer]
Hello, the following is a minimal example of training using the Mnist dataset , I hope it will help you.
https://github.com/SciSharp/TensorFlow.NET/blob/a95005fe5b0322f1a0467a7a0d61880c977f223c/test/TensorFlowNET.Keras.UnitTest/Layers/Rnn.Test.cs#L61-L79

BTW, since TensorFlow.NET has relatively few developers, its documentation is not very detailed, so I recommend you to read more test cases. ^_^

**Reply by pocky6** [reply]
Hi, 
Thanks for this minimal example. 
I saw these classes in the source of TF.Net project, but as it is slightly longer than the previous version ("keras.datasets.mnist.load_data();") I thought it was unlikely that it was the newest version, considering the documentation too (I mean, the documentation is useful and looks good, that's why I was a bit confused :) )
I'll use this one now, thank you very much :)

**Comment by Wanglongzhi2001** [comment]
BTW, I tested `keras.datasets.mnist.load_data();`, it doesn't have a problem. According to this error message`TypeLoadException : Method 'LSTMCell' in type 'Tensorflow.Keras.Layers.LayersApi' from assembly 'Tensorflow.Keras, Version=0.10.5.0, Culture=neutral, PublicKeyToken=cc7b13ffcd2ddd51' does not have an implementation.`, may I ask what is your version of TensorFlow.NET, and whether you use LSTMCell in your code?

**Reply by pocky6** [reply]
Hi again, 
Before I mark your other answer as the answer, I have to say: I'm using TF.Net 0.110.0. I don't use LSTMCell anywhere in my code.
The exception I mention was just raised by calling `keras.datasets.mnist.load_data();`.
I'll test the code from your previous answer today, if it raises the same error maybe it's a mismatch between CUDA/cudnn versions ? I have no other hypothesis about why this exception is raised.

**Reply by pocky6** [reply]
@Wanglongzhi2001 
Hi, 
I tried to call these 5 lines from my project:
```cs
 var data_loader = new MnistModelLoader(); 
 var dataset = data_loader.LoadAsync(new ModelLoadSetting 
 { 
     TrainDir = "mnist", 
     OneHot = true, 
     ValidationSize = 55000, 
 }).Result;
```
The call never returns. I tried the same from "EarlyStoppingTest" in TF.Net source, and it returns correctly. I guess something prevents TF.Net code working from my project, and it's a bit weird. Anyway, if I find more, I'll tell you.

**Reply by pocky6** [reply]
@Wanglongzhi2001 Hi,
Some news from my question.
It seems I could use these 5 lines by respecting two points: 
- not calling them from a ctor (even indirectly)
- not using .Result, but calling LoadAsync with await instead (this single change made my call successful this evening).
Anyways, thanks for your answer, and for the work on that project. I had to dive a bit into its code and found it interesting.
